exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DefaultSqlHandler.java

```diff
@@ -273,12 +282,90 @@ public RelNode visit(RelNode other) {
 
   }
 
+  /**
+   * Transform RelNode to a new RelNode without changing any traits. Also will log the outcome.
+   *
+   * @param plannerType
+   *          The type of Planner to use.
+   * @param phase
+   *          The transformation phase we're running.
+   * @param input
+   *          The origianl RelNode
+   * @return The transformed relnode.
+   */
+  private RelNode transform(PlannerType plannerType, PlannerPhase phase, RelNode input) {
+    return transform(plannerType, phase, input, input.getTraitSet());
+  }
+
+  /**
+   * Transform RelNode to a new RelNode, targeting the provided set of traits. Also will log the outcome.
+   *
+   * @param plannerType
+   *          The type of Planner to use.
+   * @param phase
+   *          The transformation phase we're running.
+   * @param input
+   *          The origianl RelNode
+   * @param targetTraits
+   *          The traits we are targeting for output.
+   * @return The transformed relnode.
+   */
+  protected RelNode transform(PlannerType plannerType, PlannerPhase phase, RelNode input, RelTraitSet targetTraits) {
+    final Stopwatch watch = Stopwatch.createStarted();
+    final RuleSet rules = config.getRules(phase);
+    final RelTraitSet toTraits = targetTraits.simplify();
+
+    final RelNode output;
+    switch (plannerType) {
+    case HEP_BOTTOM_UP:
+    case HEP: {
+      final HepProgramBuilder hepPgmBldr = new HepProgramBuilder();
+      if (plannerType == PlannerType.HEP_BOTTOM_UP) {
+        hepPgmBldr.addMatchOrder(HepMatchOrder.BOTTOM_UP);
+      }
+      for (RelOptRule rule : rules) {
+        hepPgmBldr.addRuleInstance(rule);
+      }
+
+      final HepPlanner planner = new HepPlanner(hepPgmBldr.build(), context.getPlannerSettings());
+
+      final List<RelMetadataProvider> list = Lists.newArrayList();
+      list.add(DrillDefaultRelMetadataProvider.INSTANCE);
+      planner.registerMetadataProviders(list);
+      final RelMetadataProvider cachingMetaDataProvider = new CachingRelMetadataProvider(
+          ChainedRelMetadataProvider.of(list), planner);
+
+      // Modify RelMetaProvider for every RelNode in the SQL operator Rel tree.
+      input.accept(new MetaDataProviderModifier(cachingMetaDataProvider));
+      planner.setRoot(input);
+      if (!input.getTraitSet().equals(targetTraits)) {
+        planner.changeTraits(input, toTraits);
+      }
+      output = planner.findBestExp();
+      break;
+    }
+    case VOLCANO:
+    default: {
+      // as weird as it seems, the cluster's only planner is the volcano planner.
+      final RelOptPlanner planner = input.getCluster().getPlanner();
```

Comment:
Probably better to add check if the planner is instance of Volcano. If the RelNode tree comes from DrillSqlParser.toRel() call, then it's always volcano. But what if it could come from somewhere else and it's not volcano in RelCluster.getPlanner() ?